Crash "Collection contains no element matching the predicate" in `isHasOperatorInvoke`

Hello,

the rule causes Detekt to crash.

Error:

```
> Task :app:detektDebug FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:detektDebug'.
> Analyzing .../ReportException.kt led to an exception.
  Location: com.github.kiolk.rules.UseInvokeForOperator.isHasOperatorInvoke(UseInvokeForOperator.kt:100)
  The original exception message was: Collection contains no element matching the predicate.
  Running detekt '1.23.7' on Java '17.0.12+0' on OS 'Mac OS X'
  If the exception message does not help, please feel free to create an issue on our GitHub page.
```

The problematic `ReportException.kt` file:

```kotlin
package banana

import com.google.firebase.Firebase
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.google.firebase.crashlytics.KeyValueBuilder
import com.google.firebase.crashlytics.crashlytics
import com.google.firebase.crashlytics.setCustomKeys
import timber.log.Timber
import kotlin.coroutines.cancellation.CancellationException

@Suppress("NOTHING_TO_INLINE") // inlined so Timber shows the source in tag
inline fun reportException(
    throwable: Throwable,
    message: String? = null,
    noinline customKeys: (KeyValueBuilder.() -> Unit)? = null,
) {
    Timber.e(throwable, message)
    if (throwable is CancellationException) throw throwable
    Firebase.crashlytics.recordException(throwable) {
        customKeys?.invoke(this)
        if (message != null) key("message", message)
    }
}

fun FirebaseCrashlytics.recordException(
    throwable: Throwable,
    customKeys: (KeyValueBuilder.() -> Unit)? = null,
) {
    customKeys?.let(::setCustomKeys)
    recordException(throwable)
}
```

--- a/gradle.properties
+++ b/gradle.properties
@@ -1,6 +1,6 @@
 kotlin.code.style=official
 
-versionName=1.0.5
+versionName=1.0.6
 pomGroupId=io.github.kiolk
 pomDescription=Detekt rules
 pomUrl=https://github.com/Kiolk/Detekt-rules

--- a/src/main/kotlin/com/github/kiolk/rules/UseInvokeForOperator.kt
+++ b/src/main/kotlin/com/github/kiolk/rules/UseInvokeForOperator.kt
@@ -1,23 +1,16 @@
 package com.github.kiolk.rules
 
-import io.gitlab.arturbosch.detekt.api.CodeSmell
-import io.gitlab.arturbosch.detekt.api.Config
-import io.gitlab.arturbosch.detekt.api.Debt
-import io.gitlab.arturbosch.detekt.api.Entity
-import io.gitlab.arturbosch.detekt.api.Issue
-import io.gitlab.arturbosch.detekt.api.Rule
-import io.gitlab.arturbosch.detekt.api.Severity
-import org.jetbrains.kotlin.psi.KtCallExpression
-import org.jetbrains.kotlin.psi.KtDotQualifiedExpression
-import org.jetbrains.kotlin.psi.KtExpression
-import org.jetbrains.kotlin.psi.KtPsiFactory
-import org.jetbrains.kotlin.psi.KtSafeQualifiedExpression
+import io.gitlab.arturbosch.detekt.api.*
+import io.gitlab.arturbosch.detekt.api.internal.RequiresTypeResolution
+import org.jetbrains.kotlin.incremental.components.NoLookupLocation
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.psi.*
 import org.jetbrains.kotlin.psi.psiUtil.astReplace
 import org.jetbrains.kotlin.resolve.calls.util.getType
-import org.jetbrains.kotlin.resolve.scopes.findFirstFunction
 import org.jetbrains.kotlin.types.KotlinType
 import org.jetbrains.kotlin.types.isNullable
 
+@RequiresTypeResolution
 class UseInvokeForOperator(config: Config) : Rule(config) {
 
     override val issue: Issue
@@ -31,7 +24,7 @@ class UseInvokeForOperator(config: Config) : Rule(config) {
     override fun visitDotQualifiedExpression(expression: KtDotQualifiedExpression) {
         super.visitDotQualifiedExpression(expression)
 
-        if (expression.text.contains(Regex("invoke\\(.*\\)"))) {
+        if (expression.isHasInvoke()) {
             val receivedType = expression.receiverExpression.getType(bindingContext)
 
             if (!isHasOperatorInvoke(receivedType)) return
@@ -43,7 +36,13 @@ class UseInvokeForOperator(config: Config) : Rule(config) {
             withAutoCorrect {
                 expression.astReplace(
                     KtPsiFactory.contextual(expression)
-                        .createExpression("${expression.firstChild.text}(${invokeExpression.valueArguments.joinToString(", ") { it.text }})")
+                        .createExpression(
+                            "${expression.firstChild.text}(${
+                                invokeExpression.valueArguments.joinToString(
+                                    ", "
+                                ) { it.text }
+                            })"
+                        )
                 )
             }
         }
@@ -52,7 +51,7 @@ class UseInvokeForOperator(config: Config) : Rule(config) {
     override fun visitSafeQualifiedExpression(expression: KtSafeQualifiedExpression) {
         super.visitSafeQualifiedExpression(expression)
 
-        if (expression.text.contains(Regex("invoke\\(.*\\)")) && isNotNullable(expression.receiverExpression)) {
+        if (expression.isHasInvoke() && isNotNullable(expression.receiverExpression)) {
             val receivedType = expression.receiverExpression.getType(bindingContext)
 
             if (!isHasOperatorInvoke(receivedType)) return
@@ -64,10 +63,22 @@ class UseInvokeForOperator(config: Config) : Rule(config) {
             withAutoCorrect {
                 expression.astReplace(
                     KtPsiFactory.contextual(expression)
-                        .createExpression("${expression.firstChild.text}(${invokeExpression.valueArguments.joinToString(", ") { it.text }})")
+                        .createExpression(
+                            "${expression.firstChild.text}(${
+                                invokeExpression.valueArguments.joinToString(
+                                    ", "
+                                ) { it.text }
+                            })"
+                        )
                 )
             }
         }
+    }
+
+    private fun KtQualifiedExpression.isHasInvoke(): Boolean {
+        val calledExpression = this.selectorExpression?.firstChild ?: return false
+
+        return calledExpression.text.contains(Regex("\\binvoke\\b"))
     }
 
     private fun isNotNullable(receiverExpression: KtExpression): Boolean {
@@ -87,10 +98,14 @@ class UseInvokeForOperator(config: Config) : Rule(config) {
             return false
         }
 
-        return receiverType.memberScope.findFirstFunction("invoke") { true }.isOperator
+        return receiverType.memberScope.getContributedFunctions(
+            Name.identifier("invoke"),
+            NoLookupLocation.FROM_BACKEND
+        ).firstOrNull { true }?.isOperator ?: false
     }
 
     companion object {
-        const val ISSUE_DESCRIPTION = "This rule reports that you should use operator instead using directly call invoke"
+        const val ISSUE_DESCRIPTION =
+            "This rule reports that you should use operator instead using directly call invoke"
     }
 }

--- a/src/test/kotlin/com/github/kiolk/rules/tests/UseInvokeForOperatorTest.kt
+++ b/src/test/kotlin/com/github/kiolk/rules/tests/UseInvokeForOperatorTest.kt
@@ -72,4 +72,33 @@ class UseInvokeForOperatorTest(private val environment: KotlinCoreEnvironment) {
         assertEquals(1, findings.size)
         assertEquals(UseInvokeForOperator.ISSUE_DESCRIPTION, findings[0].issue.description)
     }
+
+    @Test
+    fun `When invoke call inside of lambda which is last parameter When issue reported`() {
+        val path = compileForTest(Path("src/test/resources/InvokeCallInsideLambdaWhichLastParameter.kt"))
+
+        val findings: List<Finding> = UseInvokeForOperator(TestConfig()).compileAndLintWithContext(environment, path.text)
+
+        assertEquals(1, findings.size)
+        assertEquals(UseInvokeForOperator.ISSUE_DESCRIPTION, findings[0].issue.description)
+    }
+
+    @Test
+    fun `When invoke call inside of lambda which is last parameter of method with safe call When issue reported`() {
+        val path = compileForTest(Path("src/test/resources/SafeInvokeCallInsideLambdaWhichLastParameter.kt"))
+
+        val findings: List<Finding> = UseInvokeForOperator(TestConfig()).compileAndLintWithContext(environment, path.text)
+
+        assertEquals(1, findings.size)
+        assertEquals(UseInvokeForOperator.ISSUE_DESCRIPTION, findings[0].issue.description)
+    }
+
+    @Test
+    fun `When part of function name has invoke word When issue  is not reported`() {
+        val path = compileForTest(Path("src/test/resources/ClassWithPartMethodeNameInvoke.kt"))
+
+        val findings: List<Finding> = UseInvokeForOperator(TestConfig()).compileAndLintWithContext(environment, path.text)
+
+        assertTrue(findings.isEmpty())
+    }
 }

--- a/src/test/kotlin/com/github/kiolk/rules/utils/AutoCorrectTest.kt
+++ b/src/test/kotlin/com/github/kiolk/rules/utils/AutoCorrectTest.kt
@@ -12,7 +12,6 @@ import kotlin.test.assertEquals
 
 abstract class AutoCorrectTest<T : Rule>(factory: (config: Config) -> T, additionalConfig: List<Pair<String, Any>> = listOf()) {
 
-    private val rule: T = factory(TestConfig("autoCorrect" to false, *additionalConfig.toTypedArray()))
     private val autoCorrectRule: T = factory(TestConfig("autoCorrect" to true, *additionalConfig.toTypedArray()))
 
     fun assertLintAndFormat(

--- a/src/test/resources/ClassWithPartMethodeNameInvoke.kt
+++ b/src/test/resources/ClassWithPartMethodeNameInvoke.kt
@@ -1,0 +1,12 @@
+class ClassWithMethodeInvoke {
+
+    fun invokeMethode() {}
+}
+
+class SomeClass {
+
+    fun someFunction() {
+        val classWithMethodeInvoke = ClassWithMethodeInvoke()
+        classWithMethodeInvoke.invokeMethode()
+    }
+}

--- a/src/test/resources/InvokeCallInsideLambdaWhichLastParameter.kt
+++ b/src/test/resources/InvokeCallInsideLambdaWhichLastParameter.kt
@@ -1,0 +1,12 @@
+inline fun reportException(
+    throwable: Throwable,
+    message: String? = null,
+    noinline customKeys: (KeyValueBuilder.() -> Unit),
+) {
+    Timber.e(throwable, message)
+    if (throwable is CancellationException) throw throwable
+    Firebase.crashlytics.recordException(throwable) {
+        customKeys?.invoke()
+        if (message != null) key("message", message)
+    }
+}

--- a/src/test/resources/SafeInvokeCallInsideLambdaWhichLastParameter.kt
+++ b/src/test/resources/SafeInvokeCallInsideLambdaWhichLastParameter.kt
@@ -1,0 +1,12 @@
+inline fun reportException(
+    throwable: Throwable,
+    message: String? = null,
+    noinline customKeys: (KeyValueBuilder.() -> Unit),
+) {
+    Timber.e(throwable, message)
+    if (throwable is CancellationException) throw throwable
+    Firebase?.crashlytics?.recordException(throwable) {
+        customKeys?.invoke()
+        if (message != null) key("message", message)
+    }
+}

